### PR TITLE
feat(map): add user location features

### PR DESCRIPTION
**Pull request by MaelImhof**
# Summary

This PR introduces two new features :

1. A button to center the map view on the user's location
2. A background loop that displays the current user on the map every 5 seconds to keep it up-to-date

The permission logic is handled as well, with popups to explain the user why those permissions are needed. The app can be used without the permission.

# Details

The code of the `MapScreen` composable is becoming very hard to read and understand, I'm sorry for that. I'll see with the SCRUM master if I can be assigned to refactoring this screen's code in a future sprint, because it has become a mess.

## "Center map on me" button

In essence, this pull request has two main axes, the two features listed in the summary (center the map, update the user's location periodically).

- A "Center map on me" button was added to the map
- When you click on this button, the app checks its permissions. If it can, it will center the map on the user's location. Otherwise, it will ask for permission to access user's location.
- When asking for the permission, first a custom popup will open, explaining why the app needs the permission.
- If the user chooses to grant, the permission is requested. This logic is convoluted because of Android's permission system. Basically you have to either ask the user directly (a system popup comes into play), or bring the user to the Android settings.

## User location periodic updates

Parallel to this, the app listens to the permission state :

- When the location permission is granted, it starts periodic updates to display the up-to-date user location on the map, every five seconds
- When the location permission is revoked, it stops the periodic updates and hides the user location from the map

I also added util functions to clear the user's position, add it to a given location, and so on...

## `MapScreen` cognitive complexity

The code for the `MapScreen` composable became too complex, so I had to move some of the existing logic elsewhere to make Sonar happy.

I'm not going to refactor the entire `MapScreen.kt` file in this PR, because it is already quite significant, however I believe this is something to do, I will start the discussion on Telegram to plan such a task.

## CI

The CI does not pass yet because Sonar sees the new permission requests (coarse and fine location) as potential security issues. I believe this is the role of the reviewer to approve those issues on Sonar Cloud directly, let me know if we need to discuss it.

### CODE REVIEW

File: app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt:13]: Comment: This seems to work pretty well and I think it's good you thought about the case whenever you ask already the permission once.

File: app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt:26]: Sonar: I deem Sonar's suggestion unnecessary since the `sharedPreferences` don't hold sensitive data. So this can stay unencrypted.

File: app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt:44]: Sonar: same here.

File: app/src/main/AndroidManifest.xml
Comment (mchac15) [inline app/src/main/AndroidManifest.xml:9]: Question: Can you explain what this permissions do and why they are necessary? (Even though it seems clear I want to be sure)

**Review by mchac15 — CHANGES_REQUESTED** [review]
Overall your code is very very good.
I just have two comments:
1. You could've easily made 3 PRs with all the content there is in this one. This makes the job as a reviewer **very** difficult. I know it was handend in to you as one task, but next time please try to sepparate it into different PRs.
As you see I didn't make much comments, for me to really dig deep into all this code would take many hours which we don't have right now (I at least took 2 hours to make the PR). Fortunately you made a lot of comments and your code seems flawless. I just see a _debatable_ bug, which is one of the comments.
2. I really like the way you code it seems very well thought and the fact that you comment almost everything you do.

File: app/src/main/java/ch/hikemate/app/utils/LocationUtils.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/utils/LocationUtils.kt:36]: Fair question, I think `LocationUtils.kt` is more suited, because it is all about the location permission. In some sense, it is a special case of the `PermissionsUtils.kt`.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:187]: LGTM

File: app/src/main/java/ch/hikemate/app/utils/LocationUtils.kt
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/utils/LocationUtils.kt:36]: LGTM

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:430]: To my knowledge, it is not possible to do this easily through OSMDroid directly, so that could be possible but for another task.

File: app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt:13]: Yup, it's actually super important in Android permission management, and it took make hours to figure out lol

File: app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt:26]: Maybe you can mark the warning as accepted on SonarCloud? Or should I do this myself?

File: app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/utils/PermissionUtils.kt:44]: Maybe you can mark the warning as accepted on SonarCloud? Or should I do this myself?

File: app/src/main/AndroidManifest.xml
Comment (MaelImhof) [reply app/src/main/AndroidManifest.xml:9]: Of course. In essence, coarse location is permission to access location, but *approximative* (the precision can be terrible, it's just to get an idea of where the user is).

Fine location is more precise, but poses more problems in terms of privacy. Hence, Android guidelines tell us to, if possible, accept that the user only grants coarse location.

Since our app only displays the placement of the user of the map, I estimated it wasn't important to know the exact position of the user, so I implemented it in a way that accepts both coarse and fine location.

**Comment by MaelImhof** [comment]
Thank you for taking this review @mchac15. I agree with you, it could have been divided into 3 PRs, and I would have taken the time to do it if it had not been for the milestone deadline. Since the sensor is a requirement, I did not have the time to do it this sprint unfortunately.

I will however refactor the map screen in a future sprint, and the location logic too, to make it cleaner and more stable/maintanable.

Thank you again !

File: app/src/main/AndroidManifest.xml
Comment (mchac15) [reply app/src/main/AndroidManifest.xml:9]: Following what you are telling me I'll accept it in Sonar since it seems essential for the app's functionality.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:427]: Ok I propose it as an issue that will be solved during the refactoring of the `MapScreen`

**Review by mchac15 — APPROVED** [review]
LGTM:
The reported "bug" will be done as a task in the refractoring of the MapScreen #137

**Comment by MaelImhof** [comment]
I will wait for #115 to be merged, so I can solve the conflicts myself in this PR before merging.

**Review by mchac15 — APPROVED** [review]
LGTM:
Very good job and I hope I could help in the debuging of the PR.

**Pull request merged**
